pkg/apis/pingcap/v1alpha1/validation/validation_test.go

```diff
@@ -179,3 +181,109 @@ func TestValidateAnnotations(t *testing.T) {
 		}
 	}
 }
+
+func TestValidateNewTiDBCluster(t *testing.T) {
+	successCases := []struct {
+		name string
+		tc   v1alpha1.TidbCluster
+	}{
+		{
+			name: "all-fields-valid",
+			tc: v1alpha1.TidbCluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test",
+				},
+				Spec: v1alpha1.TidbClusterSpec{
+					Version: "v3.0.8",
+					PD: v1alpha1.PDSpec{
+						BaseImage: "pingcap/pd",
+						Config:    &v1alpha1.PDConfig{},
+						ResourceRequirements: v1.ResourceRequirements{
+							Requests: v1.ResourceList{
+								v1.ResourceStorage: resource.MustParse("10G"),
+							},
+						},
+					},
+					TiKV: v1alpha1.TiKVSpec{
+						BaseImage: "pingcap/tikv",
+						Config:    &v1alpha1.TiKVConfig{},
+						ResourceRequirements: v1.ResourceRequirements{
+							Requests: v1.ResourceList{
+								v1.ResourceStorage: resource.MustParse("10G"),
+							},
+						},
+					},
+					TiDB: v1alpha1.TiDBSpec{
+						BaseImage: "pingcap/tidb",
+						Config:    &v1alpha1.TiDBConfig{},
+					},
+				},
+			},
+		},
+	}
```

Comment:
Almost LGTM, but the unit test seems a bit of weird. I recomended to write unit test in table-driven style. FYI https://github.com/golang/go/wiki/TableDrivenTests#parallel-testing

For example, we could define testcase to determine whether the tidbcluster is valid or not, if not, which compont is going to be valid.

Here is one example. https://github.com/pingcap/tidb-operator/blob/723f7c3adda243246faac3790f917bd93b25e00e/pkg/autoscaler/autoscaler/util_test.go#L29-L94

Reply:
haha, I will fix it

Reply:
Resolved